test/service/test_manifest.py

```diff
@@ -241,6 +242,29 @@ def sort_key(row: JSON) -> bytes:
         expected.sort(key=sort_key)
         self.assertEqual(expected, manifest)
 
+    def _assert_pfb(self,
+                    expected_schema: JSON,
+                    expected_entities: JSONs,
+                    actual: Response):
+        """
+        Assert that the body of the given response contains a valid PFB manifest
+        matching the expected schema and content.
+
+        :param expected_schema: a PFB schema.
+
+        :param expected_entities: a list of PFB entities.
+
+        :param actual: an HTTP response containing a PFB manifest.
+        """
+        manifest = fastavro.reader(BytesIO(actual.content))
+        schema = manifest.writer_schema
+        # The order of the entities in the manifest is deterministic, but varies
+        # depending on the current deployment, so there's no single ordering of
```

Comment:
When using the `scan` method without `preserve_order` set to True, hits are ordered by `_doc`, i.e., the order in which they are stored in the index. We observe that each shard is considered separately when determining this order, so that every document stored at position 0 in each shard is retrieved before each document stored at position 1 in each shard, and so on. 

The difference in ordering is caused by a difference in shard count, which is in turn caused by a difference in the (non-retry) contribution concurrency.

The number of ES shards for the `replicas` index is calculated [here](https://github.com/DataBiosphere/azul/blob/de54dcb6841362843f0b90cf9bd5a251155db1a6/src/azul/indexer/index_service.py#L163).

For personal deployments, where the contribution concurrency is only 8, the shard count is determined by the ES instance count, which is 2. For other lower deployments, the shard count is 8 (64 // 8) due to the higher concurrency. On the stable deployments, I would expect a shard count of 37 (300 // 8). 

The documents appear to be assigned to and retrieved from shards deterministically (there's no direct proof of this, but the ordering never changed in the course in my testing). Therefore, we should be able to stabilize the ordering by ensuring that the unit tests are always run with a consistent shard count.

Reply:
Ordering concerns postponed to https://github.com/DataBiosphere/azul/issues/6442